onyo/lib/filters.py

```diff
@@ -0,0 +1,88 @@
+from __future__ import annotations
+import re
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from onyo.lib import Repo
+
+
+class OnyoInvalidFilterError(Exception):
+    """Raise if filters are invalidly defined"""
+
+
+def asset_name_to_keys(path: Path, default_keys: list[str]) -> dict[str, str]:
+    """Convert an asset name to default key values"""
+    return dict(zip(default_keys, re.split('[_.]', path.name)))
+
+
+@dataclass
+class Filter:
+    _arg: str = field(repr=False)
+    repo: Repo = field(repr=False)
+    key: str = field(init=False)
+    value: str = field(init=False)
+    _default_keys: list[str] = field(init=False, default_factory=list)
+
+    def __post_init__(self):
+        # TODO: define default keys somewhere accessible _once_
+        self._default_keys = ['type', 'make', 'model', 'serial']
+        self.key, self.value = self._format(self._arg)
+
+    @staticmethod
+    def _format(arg: str) -> list[str]:
+        """Split filters by the first occurrence of the `=` (equals) sign."""
+        if not isinstance(arg, str) or '=' not in arg:
+            raise OnyoInvalidFilterError(
+                'Filters must be formatted as `key=value`')
+        return arg.split('=', 1)
+
+    @property
+    def is_default(self) -> bool:
+        return True if self.key in self._default_keys else False
+
+    @staticmethod
+    def _re_match(text: str, r: str) -> bool:
+        try:
+            return True if re.compile(r).fullmatch(text) else False
+        except re.error:
+            return False
+
+    def match(self, asset: Path) -> bool:
+        """match self on asset contents which must be loaded first"""
+        unset = '<unset>'
```

Comment:
Correct, this means that `<unset>` (but also `<dict>` and `<list>`) should be defined somewhere other code can reach it. This is relevant not just for `onyo get` and `Filter`, but also for other functions like `onyo set`.

If I define these defaults in `Filter`, then even code that does not need filtering would have to access it to use defaults (and code that uses filters needs to unintuitively access it for this information). Should it be placed in `onyo/lib/__init__.py`? Anything other than that would go beyond the scope of this PR.

Reply:
Good catch. I probably refactored this 3 times before I committed, ultimately settling on a temporary `UNSET_VALUE` in `onyo/lib/filters.py`, so that slipped through the cracks. It's changed now.